### PR TITLE
Setup add widgets advanced guide

### PR DESCRIPTION
**Pull request by elenatorro**
Related issue: https://github.com/CartoDB/carto-vl/issues/1266

Guide in [staging](http://histogram-guide.developers.carto-staging.com/developers/carto-vl/guides/build-custom-charts/) 

## Context

Guide steps:

1. Simple viewportHistogram (without Airship) for category values
2. viewportHistogram for numeric values using buckets
3. viewportHistogram and globalHistogram
4. bar chart using `top`
5. adding color to a bar chart (using ramp getLegendData + histogram getJoinedValues)
6. adding color to a bar chart with `top` and with custom `others` label.

**Note** we're also saying that `top` is, for now, the only expression you can use with histograms. We're throwing an error if other expression is used in order to avoid confusion.

### CODE REVIEW

**Comment by makella** [comment]
awesome @elenatorro! 

@VictorVelarde and @rjimenezda once you have had a chance to review, let me know and I can work on the final copy editing, etc.

thanks!

**Comment by elenatorro** [comment]
Hi 👋 the main thing that has to be review from a code perspective is the changes in the `top` expression and the documentation.

Basically, at this moment `top` is the **only** "classification" expression that can be used in the histogram expressions. We'd like to improve this in the future, but for this purpose I've opened a different issue [here](https://github.com/CartoDB/carto-vl/issues/1288) that is highly related.

We've never defined a 'standard' way for `value` and `eval()`, and this is a previous step that we should tackle before moving to improve histograms to allow the usage of other classifiers.

Please take that into account (@rjimenezda and @VictorVelarde) when reviewing this issue.

File: docs/guides/12-add-widgets-advanced.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
+## Add widgets advanced
+
+We have seen how to add legends to our map by using `getLegendsData` method, and we have also learned how to display widgets using histogram expressions in the [previous guide](/developers/carto-vl/guides/add-widgets/). This is an advanced guide to learn how to obtain and display the information in widgets by using `viewportHistogram` and `globalHistogram` expressions.
+
+### Overview
+
+We will learn how to create bar charts for **categorical** data and histograms for **numeric** data. For this purpose, the following examples make use of the same source, the [**Vancouver Trees**](https://team.carto.com/u/cartovl/tables/cartovl.vancouver_trees/public/map) dataset.
```
Comment (VictorVelarde) [inline docs/guides/12-add-widgets-advanced.md]: Nice dataset. But I have seen that it has some wrong features, in our beloved [Null Island](https://en.wikipedia.org/wiki/Null_Island) at (0, 0), so we should probably delete those.

File: docs/guides/12-add-widgets-advanced.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+Bar charts and histograms have been built with [Chart.js](https://www.chartjs.org), a very simple external library. In the previous guide we used [Airship](https://carto.com/airship/) to build a histogram widget. In this case, we want to demonstrate you can use the visualization library of your choice with CARTO VL, and therefore, we have to explain how the histogram expressions work a bit deeply.
```
Comment (VictorVelarde) [inline docs/guides/12-add-widgets-advanced.md]:
```suggestion
Bar charts and histograms have been built with [Chart.js](https://www.chartjs.org), a very simple external library. In the previous guide we used [Airship](https://carto.com/airship/), which is the natural choice to build a histogram widget, but in this case we want to demonstrate how the histogram expressions work a bit deeply.
```

I would stress a bit the idea of VL <--> Airship here. That will be improved a lot with the Airship v2 VL-Bridge

File: docs/guides/12-add-widgets-advanced.md [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+### Histogram expressions
+
+There are two histogram expressions in CARTO VL: `viewportHistogram` and `globalHistogram`. These expressions return a list of values grouped by column. The `viewportHistogram` returns the list based on the features that are visible in the viewport, while `globalHistogram` takes into account the whole dataset.
```
Comment (VictorVelarde) [inline docs/guides/12-add-widgets-advanced.md]: Regarding to 'whole dataset', I think we could add here a note on the stats and how the histogram is calculated

File: docs/guides/12-add-widgets-advanced.md (hunk 2 above)
Comment (elenatorro) [reply docs/guides/12-add-widgets-advanced.md]: I agree, but I think this is not the place to do so, or maybe not at this time. Maybe we can modify this when there's a VL- Airship bridges guide, what do yo think? Explaining something that is not yet available adds a bit of noise, in my opinion.

File: docs/guides/12-add-widgets-advanced.md (hunk 3 above)
Comment (elenatorro) [reply docs/guides/12-add-widgets-advanced.md]: Agree! Good place to explain this.

File: docs/guides/12-add-widgets-advanced.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+  chart.update();
+});
+```
+
+Here you can see the result. Check how the bars change when you interact with the map.
+
+<div class="example-map">
+  <iframe
+    id="guides-widgets-advanced-step-1"
+    src="/developers/carto-vl/examples/maps/guides/add-widgets-advanced/step-1.html"
```
Comment (VictorVelarde) [inline docs/guides/12-add-widgets-advanced.md]: In this first example all the bars have the same teal color, so I guess that the previous code doesn't correspond to thist first step.

**Comment by VictorVelarde** [comment]
Nice resource @elenatorro ! 👍 
I think this will definitively help a lot when using the histograms.

I have just fixed a pair of typos and I did a little modification in the combined chart: I have configured chart.js to display both histograms in 'stacked' mode. Let me know what's your opinion on it!

**Comment by elenatorro** [comment]
Perfect! I've pending what we've discussed about removing the _null island_ and mention Airship bridges in more detail. Nice catches in the steps! And the combined example looks totally better with stacked bars. I'm going to update the staging branch so we can see there the latest changes. Nice work!

**Comment by elenatorro** [comment]
As we've talked, we should explain how the global histogram works regarding sampling. Actually, in the viewport histogram vs global histogram example, you can see there's a big difference between them:

Screenshot 2019-03-11 at 09 15 11

We should include this explanation in this section of the guide. I've already asked @jgoizueta to help us with this.

**Comment by makella** [comment]
@elenatorro will review this today :)

**Comment by makella** [comment]
hi @elenatorro I have gotten through half the guide (up to the `top` section). I mostly added information and rearranged a little bit. I will finish reviewing later today or before our meeting tomorrow.

I have a couple of questions but will share with you in our meeting tomorrow!

Feel free to accept/reject anything I did so far. I still have to re-read my own edits but wanted you to know I'm working on it!

This is going to be an awesome guide!!

**Comment by makella** [comment]
some notes from our chat:

- For the histogram chart buckets: lowest value is 0 highest value is 400+ maybe we can rearrange the breaks
- Let's see if we can explain in a little more detail the "random sampling" of global expressions so it doesn't sound so random
- think of a segway between the section for global/viewporthistograms and top()
  - Up until this point you've seen how to add two different chart types depending on the attribute as well as the differences between global and viewport based calculations. Next, let's explore how to symbolize features on the map and send that information to the charts.
- explain the map result for  step-4 with one or two sentences explaining the data/label arraysince there are labels on the chart for the first time
- think about moving `#### `viewportHistogram` vs `globalHistogram`` section to the top introduction section `## Histogram expressions`

**Comment by elenatorro** [comment]
I've changed what we talked about yesterday. About the buckets, I left them as they are for now, but we can try tomorrow better ones.

I've been testing the histograms and also trying to understand deeply how the sampling works (many thanks to @rjimenezda, @jgoizueta, @IagoLast and @rochoa). After much thought, we've decided to rename `globalHistogram` to `sampleHistogram`, at least for now, to be more realistic to what's happening behind the scenes.

As you suggested, I've moved the viewport vs **sample** histogram section to the top to explain this difference at the beginning.

We can review it together tomorrow in the meeting 👍

File: docs/guides/12-build-custom-charts.md [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+#### `viewportHistogram` vs `sampleHistogram`
+
+The map below combines both `viewportHistogram` and `sampleHistogram` expressions to compare the information returned for viewport vs sample feature calculations. If you interact with the map, you'll see how the bars for `sampleHistogram` remain static, while the ones for `viewportHistogram` change depending on the features present in the viewport.
+
+What you may notice is that if you zoom out, the `viewportHistogram` chart doesn't match the `sampleHistogram` chart. This is because the data returned for the `sampleHistogram`, as indicated by its name, is a **representative sample** of the data. Therefore, the results may vary since we're comparing the viewport data with a representative sample of the whole dataset. If you need higher accuracy, we recommend creating a [custom query](https://wiki.postgresql.org/wiki/Aggregate_Histogram) and use the [`carto.source.SQL`](/developers/carto-vl/reference/#cartosourcesql).
```
Comment (elenatorro) [inline docs/guides/12-build-custom-charts.md]: Add as a note:

> If you need higher accuracy in your `sampleHistogram`,  we recommend creating a [custom query](https://wiki.postgresql.org/wiki/Aggregate_Histogram) and use the [`carto.source.SQL`](/developers/carto-vl/reference/#cartosourcesql).

**Comment by makella** [comment]
@elenatorro 

Awesome! 

I went ahead and: 
- [x] did the copy-edits for the second half of the guide
- [x] looked through the example maps and did a few small edits with titles and symbol sizes

A couple of comments/questions:
- please read over for a final review
- do we have reference links for `viewportHistogram` and `sampleHistogram`? if so, can you link them? 
- just FYI, still seeing the labeling issue you pointed out this morning on step-6

After your final reveiw/edits, I think we're all good!

Thanks for doing this guide!

**Comment by elenatorro** [comment]
Looks great! I've added some final touches and I think we can merge it today. I've seen some issues int he documentation in the dev center while looking at staging, so I'll fix that in the branch we're I'm fixing the issues I've found while testing the examples. Nice job!! 💪

File: docs/guides/12-build-custom-charts.md [hunk 6]
```diff
@@ -166,14 +166,14 @@ You can explore this step [here](/developers/carto-vl/examples/maps/guides/build
 
 In the first example you saw how to create a chart for all category values in the street side attribute. While that attribute has four unique values, there will be other times where you may want to summarize category values in your chart based on the data. For example, if we take an attribute like tree **species name** there are many categories, but in the chart, you only want to display the **top five** tree species and their count. 
 
-You can do this using the `top` expression inside of the histogram expression: 
+You can do this using the [`top`](https://carto.com/developers/carto-vl/reference/#cartoexpressionstop) expression inside of the histogram expression:
```
Comment (VictorVelarde) [inline docs/guides/12-build-custom-charts.md]: We shouldn't use absolute urls to reference (or examples, etc.), cause then we could not support properly different versions @elenatorro @makella

Comment (elenatorro) [reply docs/guides/12-build-custom-charts.md]: You're right! Just saw it, moving it to the current PR for release fixes, nice catch!

Comment (makella) [reply docs/guides/12-build-custom-charts.md]: my fault.. did it too fast.

Comment (elenatorro) [reply docs/guides/12-build-custom-charts.md]: No problem! It's fixed ✨

**Pull request merged**